commit cda4c091a95b9753818bebb0530cfa460b5f039b
Author: Mitch Tishmack <mitch.tishmack@gmail.com>
Date:   Sun Jan 10 15:28:16 2016 -0600

    Add static flag to the stack executable

diff --git a/stack.cabal b/stack.cabal
index 9fc9a34..88311fe 100644
--- a/stack.cabal
+++ b/stack.cabal
@@ -41,6 +41,13 @@ flag disable-git-info
   -- should otherwise be avoided
   -- see: https://github.com/commercialhaskell/stack/issues/1425
 
+flag static
+  manual: True
+  default: False
+  description: Pass -static/-pthread to ghc when linking the stack binary.
+  -- Not intended for general use. Simply makes it easier to
+  -- build a fully static binary on Linux platforms that enable it.
+
 library
   hs-source-dirs:    src/
   ghc-options:       -Wall -fwarn-tabs -fwarn-incomplete-uni-patterns -fwarn-incomplete-record-updates
@@ -218,6 +225,8 @@ executable stack
   main-is:        Main.hs
   ghc-options:    -threaded -rtsopts -Wall -fwarn-tabs -fwarn-incomplete-uni-patterns -fwarn-incomplete-record-updates
   other-modules:  Paths_stack
+  if flag(static)
+      ld-options: -static -pthread
 
   build-depends:  base >=4.7 && < 5
                 , bytestring >= 0.10.4.0
